File: src/main/java/MouseEventHandler.java
import org.jetbrains.annotations.NotNull;

import java.awt.event.*;
import java.awt.geom.Point2D;

class MouseEventHandler implements MouseListener, MouseMotionListener, MouseWheelListener {
    private final Canvas canvas;
    private Point2D lastMousePosition;

    MouseEventHandler(@NotNull Canvas canvas) {
        super();
        this.canvas = canvas;
    }

    public void mouseClicked(@NotNull MouseEvent event) {
        Node node = this.canvas.getNodeUnderPoint(event.getPoint());
        if (node == null) {
            this.canvas.clearClickedNodes();
        } else {
            this.canvas.toggleClickedNode(node);
        }
    }

    public void mousePressed(@NotNull MouseEvent event) {
        this.lastMousePosition = new Point2D.Float(event.getX(), event.getY());
    }

    public void mouseReleased(@NotNull MouseEvent event) {
    }

    public void mouseEntered(@NotNull MouseEvent event) {
    }

    public void mouseExited(@NotNull MouseEvent event) {
    }

    public void mouseDragged(@NotNull MouseEvent event) {
        Point2D.Float currentMousePosition = new Point2D.Float(event.getX(), event.getY());
        if (!currentMousePosition.equals(this.lastMousePosition)) {
            Point2D currentCameraOrigin = this.canvas.getCameraOrigin();
            Point2D newCameraOrigin = new Point2D.Float(
                    (float) (currentCameraOrigin.getX() - currentMousePosition.getX() + this.lastMousePosition.getX()),
                    (float) (currentCameraOrigin.getY() - currentMousePosition.getY() + this.lastMousePosition.getY())
            );
            this.canvas
                    .setCameraOrigin(newCameraOrigin)
                    .repaint();
            this.lastMousePosition = currentMousePosition;
        }
    }

    public void mouseMoved(@NotNull MouseEvent event) {
        Node node = this.canvas.getNodeUnderPoint(event.getPoint());
        this.canvas.setHoveredNode(node);
    }

    public void mouseWheelMoved(@NotNull MouseWheelEvent event) {
        int scrollRotation = event.getWheelRotation(); // 1 if scroll down, -1 otherwise
        float zoomFactor = (float) Math.pow(1.25, -scrollRotation);
        Point2D mousePosition = new Point2D.Float(event.getX(), event.getY());
        this.canvas.zoomAtPoint(mousePosition, zoomFactor, zoomFactor);
    }
}

File: src/main/kotlin/callgraph/MouseEventHandler.kt
package callgraph

import java.awt.event.*
import java.awt.geom.Point2D

class MouseEventHandler(private val canvas: Canvas): MouseListener, MouseMotionListener, MouseWheelListener {
    private val lastMousePosition = Point2D.Float()

    override fun mouseClicked(event: MouseEvent) {
        val node = this.canvas.getNodeUnderPoint(event.point)
        if (node == null) {
            this.canvas.clearClickedNodes()
        } else {
            this.canvas.toggleClickedNode(node)
        }
    }

    override fun mousePressed(event: MouseEvent) {
        this.lastMousePosition.setLocation(event.x.toFloat(), event.y.toFloat())
    }

    override fun mouseReleased(event: MouseEvent) {
    }

    override fun mouseEntered(event: MouseEvent) {
    }

    override fun mouseExited(event: MouseEvent) {
    }

    override fun mouseDragged(event: MouseEvent) {
        val currentMousePosition = Point2D.Float(event.x.toFloat(), event.y.toFloat())
        if (currentMousePosition != this.lastMousePosition) {
            val currentCameraOrigin = this.canvas.cameraOrigin
            val newCameraOrigin = Point2D.Float(
                    currentCameraOrigin.x - currentMousePosition.x + this.lastMousePosition.x,
                    currentCameraOrigin.y - currentMousePosition.y + this.lastMousePosition.y
            )
            this.canvas.cameraOrigin.setLocation(newCameraOrigin)
            this.canvas.repaint()
            this.lastMousePosition.setLocation(currentMousePosition)
        }
    }

    override fun mouseMoved(event: MouseEvent) {
        val node = this.canvas.getNodeUnderPoint(event.point)
        this.canvas.setHoveredNode(node)
    }

    override fun mouseWheelMoved(event: MouseWheelEvent) {
        val scrollRotation = event.wheelRotation // 1 if scroll down, -1 otherwise
        val zoomFactor = Math.pow(1.25, -scrollRotation.toDouble()).toFloat()
        val mousePosition = Point2D.Float(event.x.toFloat(), event.y.toFloat())
        this.canvas.zoomAtPoint(mousePosition, zoomFactor, zoomFactor)
    }
}
